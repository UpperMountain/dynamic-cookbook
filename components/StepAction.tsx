import React from "react";
import { View, Text, TouchableOpacity } from "react-native";
import LeftLine from "./LeftLine";
import Padded from "./Padded";
import Heading from "./Heading";
import Card from "./Card";
import { padding } from "../lib/theme";
import { MaterialIcons } from "@expo/vector-icons";

function stringFormatInterval(seconds: number) {
  let minutes: number | null = null;
  let hours: number | null = null;
  if (seconds > 60) {
    minutes = Math.floor(seconds / 60);
    seconds = seconds % 60;
  }
  if (minutes && minutes > 60) {
    hours = Math.floor(minutes / 60);
    minutes = minutes % 60;
  }

  let out = "";
  if (hours) out += hours.toString().padStart(2, "0") + ":";
  if (minutes) out += minutes.toString().padStart(2, "0") + ":";
  out += seconds.toString().padStart(2, "0");

  if (!hours && !minutes) {
    out += " s";
  }

  return out;
}

export interface Props {
  // Time remaining, or undefined if not started
  remaining?: number;

  // if this card is, or leads to, a timer
  timer?: boolean;

  // "The onions are chopped"
  until: string;

  // tap callback
  onPress: () => void;
}
export default function StepAction(props: Props) {
  const { timer = false, remaining = null, until, onPress } = props;
  if (remaining !== null && !timer) {
    throw new Error("StepAction: passed time remaining without timer enabled");
  }
  return (
    <Card
      shadowAmt={0.5}
      innerStyle={{
        marginLeft: padding / 2,
        marginRight: padding / 2,
        borderRadius: 10
      }}
    >
      <TouchableOpacity
        onPress={onPress}
        style={{ marginLeft: -padding / 2, marginRight: -padding / 2 }}
      >
        <LeftLine
          hidden
          aside={
            <View
              style={{
                flex: 1,
                display: "flex",
                flexDirection: "column",
                justifyContent: "center"
              }}
            >
              <MaterialIcons
                name={
                  timer
                    ? remaining != null
                      ? "timer"
                      : "play-circle-outline"
                    : "arrow-forward"
                }
                size={40}
              />
            </View>
          }
        >
          <Padded right vertical>
            {remaining != null && (
              <Text style={{ fontSize: 30 }}>
                {stringFormatInterval(remaining)}
              </Text>
            )}
            <View
              style={{
                display: "flex",
                flexDirection: "row",
                alignItems: "center"
              }}
            >
              {remaining != null && (
                <MaterialIcons
                  name="arrow-forward"
                  size={20}
                  style={{ marginRight: 5 }}
                />
              )}
              <Heading>{until}</Heading>
            </View>
          </Padded>
        </LeftLine>
      </TouchableOpacity>
    </Card>
  );
}
