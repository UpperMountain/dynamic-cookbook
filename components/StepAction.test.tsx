import React from "react";
import StepAction from "./StepAction";
import renderer from "react-test-renderer";

describe("<StepAction/>", () => {
  it("should render without exploding", () => {
    renderer
      .create(<StepAction until="until text" onPress={jest.fn()} />)
      .unmount();
  });

  it("should pass snapshot tests", () => {
    const cmp = renderer.create(
      <>
        <StepAction until="until text" onPress={jest.fn()} />
        <StepAction
          timer={true}
          remaining={10000}
          until="until text"
          onPress={jest.fn()}
        />
        <StepAction
          timer={true}
          remaining={100}
          until="until text"
          onPress={jest.fn()}
        />
        <StepAction
          timer={true}
          remaining={10}
          until="until text"
          onPress={jest.fn()}
        />
        <StepAction timer={true} until="until text" onPress={jest.fn()} />
      </>
    );
    expect(cmp.toJSON()).toMatchSnapshot();
    cmp.unmount();
  });

  it("should throw if props make no sense", () => {
    // don't print the react error
    jest.spyOn(global.console, "error").mockImplementation(() => jest.fn());

    expect(() => {
      renderer
        .create(
          <StepAction remaining={1} until="until text" onPress={jest.fn()} />
        )
        .unmount();
    }).toThrow();

    // @ts-ignore
    console.error.mockClear();
  });
});
