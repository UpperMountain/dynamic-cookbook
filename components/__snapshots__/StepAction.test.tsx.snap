// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StepAction/> should pass snapshot tests 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "paddingBottom": 20,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "white",
            "elevation": 12,
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 5.5,
              "width": 0,
            },
            "shadowOpacity": 0.375,
            "shadowRadius": 10,
          },
          Object {
            "borderRadius": 10,
            "marginLeft": 10,
            "marginRight": 10,
          },
        ]
      }
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "marginLeft": -10,
            "marginRight": -10,
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "stretch",
                "display": "flex",
                "flexDirection": "row",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "width": 75,
                },
                Object {
                  "alignItems": "center",
                  "display": "flex",
                  "flexDirection": "column",
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "display": "flex",
                  "flex": 1,
                  "flexDirection": "column",
                  "justifyContent": "center",
                }
              }
            >
              <Text />
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "marginLeft": 0,
                },
                Object {
                  "flex": 1,
                },
                undefined,
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "paddingBottom": 20,
                    "paddingLeft": 0,
                    "paddingRight": 20,
                    "paddingTop": 20,
                  },
                  undefined,
                ]
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "display": "flex",
                    "flexDirection": "row",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "fontSize": 17,
                      },
                      undefined,
                    ]
                  }
                >
                  until text
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>,
  <View
    style={
      Array [
        Object {
          "paddingBottom": 20,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "white",
            "elevation": 12,
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 5.5,
              "width": 0,
            },
            "shadowOpacity": 0.375,
            "shadowRadius": 10,
          },
          Object {
            "borderRadius": 10,
            "marginLeft": 10,
            "marginRight": 10,
          },
        ]
      }
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "marginLeft": -10,
            "marginRight": -10,
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "stretch",
                "display": "flex",
                "flexDirection": "row",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "width": 75,
                },
                Object {
                  "alignItems": "center",
                  "display": "flex",
                  "flexDirection": "column",
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "display": "flex",
                  "flex": 1,
                  "flexDirection": "column",
                  "justifyContent": "center",
                }
              }
            >
              <Text />
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "marginLeft": 0,
                },
                Object {
                  "flex": 1,
                },
                undefined,
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "paddingBottom": 20,
                    "paddingLeft": 0,
                    "paddingRight": 20,
                    "paddingTop": 20,
                  },
                  undefined,
                ]
              }
            >
              <Text
                style={
                  Object {
                    "fontSize": 30,
                  }
                }
              >
                02:46:40
              </Text>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "display": "flex",
                    "flexDirection": "row",
                  }
                }
              >
                <Text />
                <Text
                  style={
                    Array [
                      Object {
                        "fontSize": 17,
                      },
                      undefined,
                    ]
                  }
                >
                  until text
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>,
  <View
    style={
      Array [
        Object {
          "paddingBottom": 20,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "white",
            "elevation": 12,
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 5.5,
              "width": 0,
            },
            "shadowOpacity": 0.375,
            "shadowRadius": 10,
          },
          Object {
            "borderRadius": 10,
            "marginLeft": 10,
            "marginRight": 10,
          },
        ]
      }
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "marginLeft": -10,
            "marginRight": -10,
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "stretch",
                "display": "flex",
                "flexDirection": "row",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "width": 75,
                },
                Object {
                  "alignItems": "center",
                  "display": "flex",
                  "flexDirection": "column",
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "display": "flex",
                  "flex": 1,
                  "flexDirection": "column",
                  "justifyContent": "center",
                }
              }
            >
              <Text />
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "marginLeft": 0,
                },
                Object {
                  "flex": 1,
                },
                undefined,
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "paddingBottom": 20,
                    "paddingLeft": 0,
                    "paddingRight": 20,
                    "paddingTop": 20,
                  },
                  undefined,
                ]
              }
            >
              <Text
                style={
                  Object {
                    "fontSize": 30,
                  }
                }
              >
                01:40
              </Text>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "display": "flex",
                    "flexDirection": "row",
                  }
                }
              >
                <Text />
                <Text
                  style={
                    Array [
                      Object {
                        "fontSize": 17,
                      },
                      undefined,
                    ]
                  }
                >
                  until text
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>,
  <View
    style={
      Array [
        Object {
          "paddingBottom": 20,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "white",
            "elevation": 12,
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 5.5,
              "width": 0,
            },
            "shadowOpacity": 0.375,
            "shadowRadius": 10,
          },
          Object {
            "borderRadius": 10,
            "marginLeft": 10,
            "marginRight": 10,
          },
        ]
      }
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "marginLeft": -10,
            "marginRight": -10,
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "stretch",
                "display": "flex",
                "flexDirection": "row",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "width": 75,
                },
                Object {
                  "alignItems": "center",
                  "display": "flex",
                  "flexDirection": "column",
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "display": "flex",
                  "flex": 1,
                  "flexDirection": "column",
                  "justifyContent": "center",
                }
              }
            >
              <Text />
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "marginLeft": 0,
                },
                Object {
                  "flex": 1,
                },
                undefined,
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "paddingBottom": 20,
                    "paddingLeft": 0,
                    "paddingRight": 20,
                    "paddingTop": 20,
                  },
                  undefined,
                ]
              }
            >
              <Text
                style={
                  Object {
                    "fontSize": 30,
                  }
                }
              >
                10 s
              </Text>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "display": "flex",
                    "flexDirection": "row",
                  }
                }
              >
                <Text />
                <Text
                  style={
                    Array [
                      Object {
                        "fontSize": 17,
                      },
                      undefined,
                    ]
                  }
                >
                  until text
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>,
  <View
    style={
      Array [
        Object {
          "paddingBottom": 20,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "white",
            "elevation": 12,
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 5.5,
              "width": 0,
            },
            "shadowOpacity": 0.375,
            "shadowRadius": 10,
          },
          Object {
            "borderRadius": 10,
            "marginLeft": 10,
            "marginRight": 10,
          },
        ]
      }
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "marginLeft": -10,
            "marginRight": -10,
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "stretch",
                "display": "flex",
                "flexDirection": "row",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "width": 75,
                },
                Object {
                  "alignItems": "center",
                  "display": "flex",
                  "flexDirection": "column",
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "display": "flex",
                  "flex": 1,
                  "flexDirection": "column",
                  "justifyContent": "center",
                }
              }
            >
              <Text />
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "marginLeft": 0,
                },
                Object {
                  "flex": 1,
                },
                undefined,
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "paddingBottom": 20,
                    "paddingLeft": 0,
                    "paddingRight": 20,
                    "paddingTop": 20,
                  },
                  undefined,
                ]
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "display": "flex",
                    "flexDirection": "row",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "fontSize": 17,
                      },
                      undefined,
                    ]
                  }
                >
                  until text
                </Text>
              </View>
              <Text>
                Start timer
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>,
]
`;
